### PR TITLE
Import a file to a disk

### PR DESCRIPTION
**Pull request by jmpesp**
Use the bulk-write import functionality to upload a file to a disk,
optionally making a snapshot and an image out of that snapshot. An
example invocation:

    oxide disk import \
        --path crucible-tester-sparse.img \
        --threads 8 \
        --disk-name crucible-tester-disk \
        --project myproj \
        --description "testing import!" \
        --snapshot-name crucible-tester-snapshot \
        --image-name crucible-tester \
        --image-description "uploading of crucible-tester-sparse.img" \
        --image-os debian \
        --image-version 11

This will:

- create a disk named `crucible-tester-disk` in state "import-ready"
- set that disk to "importing-via-bulk-write"
- spawn 8 upload tasks and bulk write file chunks to the disk in
  parallel
- wait for it to be done
- stop the bulk write process
- finalize the disk, creating the snapshot `crucible-tester-snapshot`
- create the image `crucible-tester` out of that snapshot

### CODE REVIEW

**Comment by jmpesp** [comment]
Closes https://github.com/oxidecomputer/oxide.rs/issues/82

File: cli/src/cmd_disk.rs [hunk 1]
```diff
@@ -4,27 +4,457 @@
...
+        rounded_down_gb + ONE_GB
+    } else {
+        disk_size
+    };
+
+    Ok(disk_size)
+}
+
+// Upload to Nexus in 512k byte chunks
+const CHUNK_SIZE: u64 = 512 * 1024;
```
Comment (david-crespo) [inline cli/src/cmd_disk.rs:176]: This seems really small!

Comment (david-crespo) [reply cli/src/cmd_disk.rs:176]: Ah. It comes from Crucible: https://github.com/oxidecomputer/crucible/blob/0b34a53b57038a7804ac69ab90400539bd1961e8/pantry/src/pantry.rs#L33

File: cli/src/cmd_disk.rs [hunk 2]
```diff
@@ -4,27 +4,457 @@
...
+            let pb = mb.add(ProgressBar::new(disk_size / self.threads as u64));
+            pb.set_style(ProgressStyle::default_bar()
+                .template("[{elapsed_precise}] [{wide_bar:.green}] {bytes}/{total_bytes} ({bytes_per_sec}, {eta})")?
+            );
+
+            let client = client.clone();
+            let disk_name = self.disk_name.clone();
+            let project = self.project.clone();
+
+            handles.push(tokio::spawn(async move {
```
Comment (ahl) [inline cli/src/cmd_disk.rs]: I guess a task is the same as a thread to you...

**Comment by jmpesp** [comment]
@ahl ready for review. I've done what tests I could given that httpmock can't return different responses for the same request.

File: cli/src/cmd_disk.rs [hunk 3]
```diff
@@ -4,27 +4,518 @@
...
+    description: String,
+
+    /// Path to the file to import
+    #[clap(long)]
+    path: String,
+
+    /// How many upload threads to run. Disk size must be divisible by this
+    /// number.
+    #[clap(long, default_value = "1")]
+    threads: usize,
```
Comment (ahl) [inline cli/src/cmd_disk.rs]: This seems challenging for users; can we eliminate this constraint?

Can we rename this `concurrency`? Why would someone change this from the default? Presumably to make the data transfer faster. How would I know if the right value were 2 or 20 or 200? Is this particularly CPU intensive? That seems unlikely. Am I really trying to influence bandwidth consuption?

File: cli/src/cmd_disk.rs [hunk 4]
```diff
@@ -4,27 +4,518 @@
...
+
+    /// The name of the disk to create
+    #[clap(long)]
+    disk_name: Name,
+
+    /// The size of the disk to create. If unspecified, the size of the file
+    /// will be used, rounded up to the nearest GB.
+    #[clap(long)]
+    disk_size: Option<u64>,
```
Comment (ahl) [inline cli/src/cmd_disk.rs]: I **think** if you use `oxide_api::types::ByteCount` you'll get suffix parsing "for free"

File: cli/src/cmd_disk.rs [hunk 5]
```diff
@@ -4,27 +4,518 @@
...
+
+    /// The desired size of the disk's blocks.
+    #[clap(long)]
+    disk_block_size: Option<i64>,
```
Comment (ahl) [inline cli/src/cmd_disk.rs]: I see that we default to 512 below (is that what we want?). We might want to use `oxide_api::types::BlockSize`. I'm not exactly sure of the best way to do this, but it's possible to have clap show us only the allowable values for this property in the help output. I think this may involve either adding your own enum or adding a new impl to `clap_feature.rs` for `BlockSize`

File: cli/src/cmd_disk.rs [hunk 6]
```diff
@@ -4,27 +4,518 @@
...
+    /// If supplied, create an image from the snapshot of this disk. Requires
+    /// snapshot name to be supplied.
+    #[clap(long)]
+    image_name: Option<Name>,
+
+    /// The description for the image created out of the snapshot of this disk.
+    #[clap(long)]
+    image_description: Option<String>,
+
+    /// The OS of this image (eg. Debian)
```
Comment (ahl) [inline cli/src/cmd_disk.rs]:
```suggestion
    /// The OS of this image (e.g. Debian)
```

File: cli/src/cmd_disk.rs [hunk 7]
```diff
@@ -4,27 +4,518 @@
...
+    #[clap(long)]
+    image_os: Option<String>,
+
+    /// The version of this image (eg. 11, focal, a9e77e3a, 2023-04-06T14:23:34Z)
```
Comment (ahl) [inline cli/src/cmd_disk.rs]:
```suggestion
    /// The version of this image (e.g. 11, focal, a9e77e3a, 2023-04-06T14:23:34Z)
```

File: cli/src/cmd_disk.rs [hunk 8]
```diff
@@ -4,27 +4,518 @@
...
+use oxide_api::ClientDisksExt;
+use oxide_api::ClientImagesExt;
+use oxide_api::ClientSnapshotsExt;
+use oxide_api::ResponseValue;
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+use tokio::sync::mpsc;
 
+/// Import a file to a disk
```
Comment (ahl) [inline cli/src/cmd_disk.rs]:
```suggestion
/// Create a disk from a file upload
```

maybe?

File: cli/src/cmd_disk.rs [hunk 9]
```diff
@@ -4,27 +4,518 @@
...
+///
+/// Upload the contents of a file to a disk, optionally creating a snapshot
+/// and/or an image.
```
Comment (ahl) [inline cli/src/cmd_disk.rs]:
```suggestion
/// Upload the contents of a file to create a new disk. Use the --snapshot
/// or --image options to also create a snapshot or image.
```

again: maybe?

File: cli/src/cmd_disk.rs [hunk 10]
```diff
@@ -4,27 +4,518 @@
...
+    #[clap(long)]
+    disk_size: Option<u64>,
+
+    /// The desired size of the disk's blocks.
+    #[clap(long)]
+    disk_block_size: Option<i64>,
+
+    /// If supplied, create a snapshot with this name.
+    #[clap(long)]
+    snapshot_name: Option<Name>,
```
Comment (ahl) [inline cli/src/cmd_disk.rs]:
```suggestion

    /// If supplied, create a snapshot with the given name.
    #[clap(long)]
    snapshot: Option<Name>,
```

File: cli/src/cmd_disk.rs [hunk 11]
```diff
@@ -4,27 +4,518 @@
...
+
+    /// If supplied, create an image from the snapshot of this disk. Requires
+    /// snapshot name to be supplied.
+    #[clap(long)]
+    image_name: Option<Name>,
```
Comment (ahl) [inline cli/src/cmd_disk.rs]:
```suggestion
    /// If supplied, create an image with the given name. Requires the creation of a snapshot.
    #[clap(long, requires = "snapshot")]
    image: Option<Name>,
```

File: cli/src/cmd_disk.rs [hunk 12]
```diff
@@ -4,27 +4,518 @@
...
+
+    /// The description for the image created out of the snapshot of this disk.
+    #[clap(long)]
```
Comment (ahl) [inline cli/src/cmd_disk.rs]:
```suggestion
    #[clap(long, requires = "snapshot", requires = "image")]
```

File: cli/src/cmd_disk.rs [hunk 13]
```diff
@@ -4,27 +4,518 @@
...
+    /// The OS of this image (eg. Debian)
+    #[clap(long)]
+    image_os: Option<String>,
+
+    /// The version of this image (eg. 11, focal, a9e77e3a, 2023-04-06T14:23:34Z)
+    #[clap(long)]
+    image_version: Option<String>,
+}
+
+/// Return a correct disk size for the path and size arguments
```
Comment (ahl) [inline cli/src/cmd_disk.rs]: What does "correct" mean in this context?

File: cli/src/cmd_disk.rs [hunk 14]
```diff
@@ -4,27 +4,518 @@
...
+    const ONE_GB: u64 = 1024 * 1024 * 1024;
+
+    let disk_size = if let Some(size) = size {
+        size
+    } else {
+        std::fs::metadata(path)?.len()
+    };
+
+    // Nexus' disk size minimum is 1 GB
+    let disk_size = std::cmp::max(ONE_GB, disk_size);
```
Comment (ahl) [inline cli/src/cmd_disk.rs]: is this needed given the code below that rounds up?

File: cli/src/cmd_disk.rs [hunk 15]
```diff
@@ -4,27 +4,518 @@
...
+
+            // The finalize succeeded, so delete the disk.
+            self.unwind_disk_delete(&client).await?;
+
+            // Finalizing and deleting the disk succeeded, so return the
+            // original error.
+            return Err(e.into());
+        }
+
+        // Create one tokio task for each thread that will upload file chunks
```
Comment (ahl) [inline cli/src/cmd_disk.rs:342]: would this be simpler with rayon... or more complicated?

File: cli/src/cmd_disk.rs [hunk 16]
```diff
@@ -4,27 +4,518 @@
...
+            .disk(self.disk_name.clone())
+            .send()
+            .await;
+
+        if let Err(e) = stop_bulk_write_response {
+            eprintln!("stopping the bulk write process failed with {:?}", e);
+
+            // Attempt to unwind the disk, although it will probably fail - the
+            // first step is to stop the bulk write process!
+            self.unwind_disk_bulk_write_stop(&client).await?;
```
Comment (ahl) [inline cli/src/cmd_disk.rs:469]: might we want to retry for a bit? Perhaps depending on the status code?

File: cli/src/cmd_disk.rs [hunk 17]
```diff
@@ -4,27 +4,518 @@
...
+            return Err(e.into());
+        }
+
+        // Finalize the disk, optionally making a snapshot
+        let request = client
+            .disk_finalize_import()
+            .project(&self.project)
+            .disk(self.disk_name.clone())
+            .body(FinalizeDisk {
+                snapshot_name: self.snapshot_name.as_ref().map(|x| x.clone()),
```
Comment (ahl) [inline cli/src/cmd_disk.rs]:
```suggestion
                snapshot_name: self.snapshot_name.clone(),
```
?

File: cli/src/cmd_disk.rs (hunk 12 above)
Comment (ahl) [reply cli/src/cmd_disk.rs]: It seems like image/image-name requires image-description, image-os, and image-version as well, yes? You could make a clap group to encode this.

File: cli/src/cmd_disk.rs [hunk 18]
```diff
@@ -4,27 +4,518 @@
...
+            // a snapshot out of it. if this step fails, they can always
+            // manually make an image out of the snapshot later and be sure that
+            // the snapshot's contents are the same.
+            let snapshot: Snapshot = client
+                .snapshot_view()
+                .project(&self.project)
+                .snapshot(NameOrId::Name(snapshot_name.clone()))
+                .send()
+                .await?
+                .into_inner();
```
Comment (ahl) [inline cli/src/cmd_disk.rs]: for my education: are the explicit type and `into_inner()` call necessary here?

File: cli/src/main.rs [hunk 19]
```diff
@@ -137,7 +137,8 @@ async fn main() {
 
         // TODO this could be improved
```
Comment (ahl) [inline cli/src/main.rs:138]:
```suggestion
```

File: cli/tests/test_disk_import.rs [hunk 20]
```diff
@@ -0,0 +1,550 @@
...
+        .arg("--project")
+        .arg("myproj")
+        .arg("--description")
+        .arg("disk description")
+        .arg("--path")
+        .arg("tests/data/testpost.iso")
+        .arg("--disk-name")
+        .arg("test-import")
+        .assert()
+        .success();
```
Comment (ahl) [inline cli/tests/test_disk_import.rs:156]: do you want to look at the command output? No is a fine answer ;-)

File: cli/tests/test_disk_import.rs [hunk 21]
```diff
@@ -0,0 +1,550 @@
...
+        .arg("--snapshot-name")
+        .arg("test-import-snap")
+        .arg("--image-name")
+        .arg("test-import")
+        .arg("--image-description")
+        .arg("image description")
+        .arg("--image-os")
+        .arg("dos")
+        .arg("--image-version")
+        .arg("6.1")
```
Comment (ahl) [inline cli/tests/test_disk_import.rs:284]: makes sense.

File: cli/tests/test_disk_import.rs [hunk 22]
```diff
@@ -0,0 +1,550 @@
...
+    let server = MockServer::start();
+
+    let disk_view_mock = server.disk_view(|when, then| {
+        when.into_inner().any_request();
+        then.client_error(
+            404,
+            &oxide_api::types::Error {
+                error_code: None,
+                message: "disk not found".into(),
+                request_id: Uuid::new_v4().to_string(),
```
Comment (ahl) [inline cli/tests/test_disk_import.rs]: If you want this to be predictable you **could** do `Uuid::mock_value(&mut src)`

File: test-common/src/json_mock.rs [hunk 23]
```diff
@@ -176,6 +176,24 @@ fn mock_schema_object(
             extensions: _,
         } if instance_type.as_ref() == &InstanceType::String => mock_string(format, string, src),
 
+        // UUID
```
Comment (ahl) [inline test-common/src/json_mock.rs]:
```suggestion
        // One way to express an option
```

File: test-common/src/json_mock.rs [hunk 24]
```diff
@@ -176,6 +176,24 @@ fn mock_schema_object(
...
+            enum_values: None,
+            const_value: None,
+            subschemas: None,
+            number: _,
+            string,
+            array: _,
+            object: _,
+            reference: None,
+            extensions: _,
+        } if matches!(&vec[..], &[InstanceType::String, InstanceType::Null]) => {
```
Comment (ahl) [inline test-common/src/json_mock.rs]: not sure this is strictly necessary, but for robustness:

```suggestion
        } if matches!(&vec[..], &[InstanceType::String, InstanceType::Null] | &[InstanceType::Null, InstanceType::String]) => {
```

Not required here, but I think I'll change this in the future to do something like we do in typify:

https://github.com/oxidecomputer/typify/blob/main/typify-impl/src/convert.rs#L56

File: test-common/src/json_mock.rs [hunk 25]
```diff
@@ -233,6 +252,22 @@ fn mock_schema_object(
             mock_schema(schema, definitions, src)
         }
 
+        // one way that Option is generated
```
Comment (ahl) [inline test-common/src/json_mock.rs]:
```suggestion
        // The null schema--we see this sometimes in an `anyOf` resulting from an `Option`
```

**Review by ahl — COMMENTED** [review]
looks good; some notes regarding args and arg parsing

File: cli/src/cmd_disk.rs (hunk 3 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: I basically always set this to 8. Each of the upload tasks goes at about 5 MiB/s, so the aggregate is 40 MiB/s. Above 8 tasks, the aggregate doesn't go up. Maybe it'd be better to have one progress bar for all upload threads, set this to 8 by default, and rename it `upload-concurrency`?

I'll stare at the code a bit to see if I can remove the divisible constraint.

File: cli/src/cmd_disk.rs (hunk 4 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: Done in d8837c8

File: cli/src/cmd_disk.rs (hunk 5 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: I've done this in d8837c8, but it looks like the error messages could be better:

```
$ ./cli.sh disk import --disk-size 123a
error: one of the values isn't valid for an argument
$ ./cli.sh disk import --disk-block-size 123
error: one of the values isn't valid for an argument
```

File: cli/src/cmd_disk.rs (hunk 6 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: done in 3d5bdd8e6e968bfee866985c38ebae5044521e5f

File: cli/src/cmd_disk.rs (hunk 7 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: done in 3d5bdd8e6e968bfee866985c38ebae5044521e5f

File: cli/src/cmd_disk.rs (hunk 8 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: changed in 3d5bdd8e6e968bfee866985c38ebae5044521e5f - let me know what you think

File: cli/src/cmd_disk.rs (hunk 9 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: changed in 3d5bdd8e6e968bfee866985c38ebae5044521e5f - let me know what you think

File: cli/src/cmd_disk.rs (hunk 10 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: changed in 3d5bdd8e6e968bfee866985c38ebae5044521e5f

File: cli/src/cmd_disk.rs (hunk 11 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: changed in 3d5bdd8e6e968bfee866985c38ebae5044521e5f

File: cli/src/cmd_disk.rs (hunk 13 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: changed in 3d5bdd8e6e968bfee866985c38ebae5044521e5f

File: cli/src/cmd_disk.rs (hunk 14 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: no, changed in ffab026

File: cli/src/cmd_disk.rs (hunk 17 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: cool! changed in 1670a86

File: cli/src/main.rs (hunk 19 above)
Comment (jmpesp) [reply cli/src/main.rs:138]: turns out no! ff59b45

File: cli/tests/test_disk_import.rs (hunk 20 above)
Comment (jmpesp) [reply cli/tests/test_disk_import.rs:156]: not in this case :)

File: cli/tests/test_disk_import.rs (hunk 22 above)
Comment (jmpesp) [reply cli/tests/test_disk_import.rs]: I like predictable, changed in 8d43e6d

File: test-common/src/json_mock.rs (hunk 23 above)
Comment (jmpesp) [reply test-common/src/json_mock.rs]: done in f393d4e

File: test-common/src/json_mock.rs (hunk 24 above)
Comment (jmpesp) [reply test-common/src/json_mock.rs]: done in f393d4e

File: test-common/src/json_mock.rs (hunk 25 above)
Comment (jmpesp) [reply test-common/src/json_mock.rs]: done in f393d4e

File: cli/src/cmd_disk.rs (hunk 3 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: 3433f6675cc1861d11be9716f8b81339d8ce7bc5 does away with the `--threads` argument, hard coding it to 8. it also loosens the constraint for the user - their file size must be evenly divided by the block size. The pantry will reject writes where this isn't true anyway.

File: sdk/src/clap_feature.rs [hunk 26]
```diff
@@ -58,3 +58,37 @@ impl clap::builder::TypedValueParser for ByteCountParser {
...
+        ) -> Option<types::BlockSize> {
+            let value: Option<i64> = value.to_str()?.parse().ok();
+            match value {
+                Some(v) => v.try_into().ok(),
+                None => None,
+            }
+        }
+
+        parse(cmd, arg, value).ok_or_else(|| clap::Error::new(clap::error::ErrorKind::InvalidValue))
+    }
```
Comment (ahl) [inline sdk/src/clap_feature.rs]: why was this needed?

File: cli/src/cmd_disk.rs (hunk 12 above)
Comment (ahl) [reply cli/src/cmd_disk.rs]: still this? ^

File: cli/src/cmd_disk.rs [hunk 27]
```diff
@@ -4,27 +4,550 @@
...
+                "file size {} is not divisible by block size{}!",
+                file_size,
+                *disk_block_size
+            );
+        }
+
+        // Use 8 upload tasks - this evenly divides all block sizes, and we know
+        // that the file size is evenly divided by the selected block size due
+        // to the above check.
+        const UPLOAD_TASKS: usize = 8;
```
Comment (ahl) [inline cli/src/cmd_disk.rs:302]: love it

**Review by ahl — COMMENTED** [review]
looks good to me

File: sdk/src/clap_feature.rs (hunk 26 above)
Comment (jmpesp) [reply sdk/src/clap_feature.rs]: Otherwise, I get
```
error[E0599]: the method `value_parser` exists for reference `&&&&&&_AutoValueParser<BlockSize>`, but its trait bounds were not satisfied
    --> cli/src/cmd_disk.rs:63:5
     |
63   |     /// The desired size of the disk's blocks.
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method cannot be called on `&&&&&&_AutoValueParser<BlockSize>` due to unsatisfied trait bounds
     |
    ::: /home/jwm/.cargo/registry/src/github.com-1ecc6299db9ec823/clap_builder-4.2.4/src/builder/value_parser.rs:2221:1
     |
2221 | pub struct _AutoValueParser<T>(std::marker::PhantomData<T>);
     | ------------------------------ doesn't satisfy `_: _ValueParserViaParse`
     |
    ::: /home/jwm/src/oxidecomputer/oxide.rs/sdk/src/generated_sdk.rs:127:5
     |
127  |     pub struct BlockSize(i64);
     |     --------------------
     |     |
     |     doesn't satisfy `BlockSize: From<&'s std::ffi::OsStr>`
     |     doesn't satisfy `BlockSize: From<&'s str>`
     |     doesn't satisfy `BlockSize: From<OsString>`
     |     doesn't satisfy `BlockSize: From<std::string::String>`
     |     doesn't satisfy `BlockSize: FromStr`
     |     doesn't satisfy `BlockSize: ValueEnum`
     |     doesn't satisfy `BlockSize: ValueParserFactory`
     |
     = note: the following trait bounds were not satisfied:
             `BlockSize: ValueEnum`
             which is required by `&&&&&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaValueEnum`
             `BlockSize: ValueParserFactory`
             which is required by `&&&&&&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaFactory`
             `BlockSize: From<OsString>`
             which is required by `&&&&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaFromOsString`
             `BlockSize: From<&'s std::ffi::OsStr>`
             which is required by `&&&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaFromOsStr`
             `BlockSize: From<std::string::String>`
             which is required by `&&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaFromString`
             `BlockSize: From<&'s str>`
             which is required by `&_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaFromStr`
             `BlockSize: FromStr`
             which is required by `_AutoValueParser<BlockSize>: clap::builder::via_prelude::_ValueParserViaParse`
     = note: this error originates in the macro `clap::value_parser` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0599`.
error: could not compile `oxide` due to 2 previous errors
```

File: cli/src/cmd_disk.rs (hunk 27 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs:302]: :+1:

File: sdk/src/clap_feature.rs (hunk 26 above)
Comment (ahl) [reply sdk/src/clap_feature.rs]: ok; you might want to just do the validation inline: "512", "2048", "2k" ... (not sure why we support 2k, but I'm sure there are reasons...)

File: cli/src/cmd_disk.rs (hunk 12 above)
Comment (jmpesp) [reply cli/src/cmd_disk.rs]: see "I wasn't smart enough to get groups to work" @ 47ef018 - I used explicitly provided `requires_all`

File: sdk/src/clap_feature.rs (hunk 26 above)
Comment (jmpesp) [reply sdk/src/clap_feature.rs]: > do the validation inline

I'm not 100% sure what you mean by this but my assumption is

```patch
diff --git a/sdk/src/clap_feature.rs b/sdk/src/clap_feature.rs
index 04dfc65..9348fa7 100644
--- a/sdk/src/clap_feature.rs
+++ b/sdk/src/clap_feature.rs
@@ -84,11 +84,16 @@ impl clap::builder::TypedValueParser for BlockSizeParser {
         ) -> Option<types::BlockSize> {
             let value: Option<i64> = value.to_str()?.parse().ok();
             match value {
-                Some(v) => v.try_into().ok(),
+                Some(512) => 512.try_into().ok(),
+                Some(2048) => 2048.try_into().ok(),
+                Some(4096) => 4096.try_into().ok(),
+                Some(_) => None,
                 None => None,
             }
         }
```

which I've committed (along with a better error message) in 6847f20

**Comment by jmpesp** [comment]
I'm able to reproduce this test failure on my workstation, but it's intermittent!

```
---- test_disk_import_bulk_write_import_fail stdout ----
thread 'test_disk_import_bulk_write_import_fail' panicked at 'assertion failed: `(left == right)`
  left: `"^/v1/disks/.*/bulk-write$"`,
 right: `"/v1/disks/test-import"`: 1 of 2 expected requests matched the mock specification, .
Here is a comparison with the most similar non-matching request (request number 1): 

1 : The path does not match
------------------------------------------------------------------------------------------
Expected:	[matches regex]		^/v1/disks/.*/bulk-write$
Actual:		                    	/v1/disks/test-import

2 : The method does not match
------------------------------------------------------------------------------------------
Expected:	[equals]		POST
Actual:		             	GET

jwm@duster ~/src/oxidecomputer/oxide.rs [image_import] [] $ cargo test --all-features
    Finished test [unoptimized + debuginfo] target(s) in 0.40s
     Running unittests src/main.rs (target/debug/deps/oxide-befbc3f1a577763c)

running 6 tests
test cmd_auth::test::test_parse_host ... ok
test tests::test_autoref ... ok
test cmd_api::test::test_add_query_string ... ok
test cmd_disk::test_get_disk_size ... ok
test cmd_version::version_success ... ok
test cmd_auth::test_cmd_auth_status ... ok

test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.54s

     Running tests/test_api.rs (target/debug/deps/test_api-24c40ef8fd905d65)

running 3 tests
test test_pagination_midway_failure ... ok
test test_pagination_success ... ok
test test_simple_api_call ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.09s

     Running tests/test_disk_import.rs (target/debug/deps/test_disk_import-0bebbfea6cf5888d)

running 10 tests
test test_disk_import_bad_file_size ... ok
test test_disk_import_bad_block_size ... ok
test test_disk_import_disk_exists_already ... ok
test test_disk_import_bulk_import_start_fail ... ok
test test_disk_import_image_exists_already ... ok
test test_disk_import_required_parameters ... ok
test test_disk_import_snapshot_exists_already ... ok
test test_disk_import_sparse ... ok
test test_disk_import_bulk_write_import_fail ... ok
test test_disk_import ... ok

...
```

File: cli/tests/test_disk_import.rs [hunk 28]
```diff
@@ -0,0 +1,772 @@
...
+    disk_bulk_write_mock.assert_hits(2);
+    unwind_stop_bulk_write_mock.assert();
+    unwind_finalize_mock.assert();
+    unwind_disk_delete_mock.assert();
+}
+
+// A disk import where the requested block size is invalid
+#[test]
+fn test_disk_import_bad_block_size() {
+    let server = MockServer::start();
```
Comment (ahl) [inline cli/tests/test_disk_import.rs]: unneeded?

File: cli/tests/test_disk_import.rs [hunk 29]
```diff
@@ -0,0 +1,772 @@
...
+        .arg("--disk")
+        .arg("test-import")
+        .assert()
+        .failure();
+}
+
+// A disk import where the file size doesn't divide the block size
+#[test]
+fn test_disk_import_bad_file_size() {
+    let server = MockServer::start();
```
Comment (ahl) [inline cli/tests/test_disk_import.rs]: unneeded?

File: cli/tests/test_disk_import.rs [hunk 30]
```diff
@@ -0,0 +1,772 @@
...
+        .arg("--disk")
+        .arg("test-import")
+        .assert()
+        .failure();
+}
+
+// Test for required parameters being supplied
+#[test]
+fn test_disk_import_required_parameters() {
+    let server = MockServer::start();
```
Comment (ahl) [inline cli/tests/test_disk_import.rs]: unneeded?

**Review by ahl — APPROVED** [review]
I approve per your recent threats.
